### PR TITLE
FEC-11577 AdLayout Configuration

### PR DESCRIPTION
**Pull request by GouravSna**
### Description of the Changes

[Works with [Kaltura-Player](https://github.com/kaltura/kaltura-player-android)]

With Ad layout config you can create your own ad break timeline using your vast tags. Ad break can be set as pre, mid and post rolls and each ad break can contain a single vast tag or multiple tags, either as a pod, but also as a Waterfall.

### CODE REVIEW

File: imaplugin/src/main/java/com/kaltura/playkit/plugins/ima/IMAPlugin.java [hunk 1]
```diff
@@ -608,25 +617,34 @@ private void cancelAdManagerTimer() {
         }
     }
 
-    private void requestAdsFromIMA(String adTagUrl) {
+    private void requestAdsOnUpdateMedia(String adTagUrl) {
+        log.d("Do requestAdsOnChangeMedia");
```
Comment (giladna) [inline imaplugin/src/main/java/com/kaltura/playkit/plugins/ima/IMAPlugin.java]: @GouravSna  please fix log 
requestAdsOnUpdateMedia

File: imaplugin/src/main/java/com/kaltura/playkit/plugins/ima/IMAPlugin.java [hunk 2]
```diff
@@ -775,6 +806,52 @@ public void contentCompleted() {
...
+        if (isAdvertisingConfigured) {
+            isAdvertisingConfigLoading = !TextUtils.isEmpty(adTagUrl);
+            requestAdsFromIMA(null, adTagUrl);
+        }
+    }
+
+    @Override
+    public void setAdvertisingConfig(boolean isConfigured, IMAEventsListener imaEventsListener) {
+        isAdvertisingConfigured = isConfigured;
+        this.imaEventsListener = imaEventsListener;
```
Comment (giladna) [inline imaplugin/src/main/java/com/kaltura/playkit/plugins/ima/IMAPlugin.java:855]: @GouravSna  where do you remove this listener when done?

File: imaplugin/src/main/java/com/kaltura/playkit/plugins/ima/IMAPlugin.java [hunk 3]
```diff
@@ -775,6 +806,52 @@ public void contentCompleted() {
...
+    }
+
+    @Override
+    public void setCuePoints(List<Long> cuePoints, AdBreakPositionType adBreakPositionType, boolean isUpdatedCuePoint) {
+        if (isAdvertisingConfigured && cuePoints != null && !cuePoints.isEmpty()) {
+            advertisingConfigCuePoints = cuePoints;
```
Comment (giladna) [inline imaplugin/src/main/java/com/kaltura/playkit/plugins/ima/IMAPlugin.java:863]: @GouravSna do you clean it up for change media?

Comment (GouravSna) [reply imaplugin/src/main/java/com/kaltura/playkit/plugins/ima/IMAPlugin.java:863]: Fixed

File: imaplugin/src/main/java/com/kaltura/playkit/plugins/ima/IMAPlugin.java (hunk 2 above)
Comment (GouravSna) [reply imaplugin/src/main/java/com/kaltura/playkit/plugins/ima/IMAPlugin.java:855]: Fixed

File: imaplugin/src/main/java/com/kaltura/playkit/plugins/ima/IMAPlugin.java [hunk 4]
```diff
@@ -678,6 +722,7 @@ public void onFinish() {
                     if (getPlayerEngine() != null) {
                         getPlayerEngine().play();
                     }
+                    //TODO: Check if content player should be hidden here
```
Comment (giladna) [inline imaplugin/src/main/java/com/kaltura/playkit/plugins/ima/IMAPlugin.java]: what is it?

File: imaplugin/src/main/java/com/kaltura/playkit/plugins/ima/IMAPlugin.java [hunk 5]
```diff
@@ -608,25 +631,34 @@ private void cancelAdManagerTimer() {
...
         }
         resetIMA();
 
-        log.d("Do requestAdsFromIMA");
+        if (!isAdvertisingConfigured) {
+            requestAdsFromIMA(adTagResponse, adTagUrl);
+        }
+    }
+
+    private void requestAdsFromIMA(String adTagResponse, String adTagUrl) {
```
Comment (giladna) [inline imaplugin/src/main/java/com/kaltura/playkit/plugins/ima/IMAPlugin.java:658]: @GouravSna  I think ad tag should be first parameter

File: imaplugin/src/main/java/com/kaltura/playkit/plugins/ima/IMAPlugin.java [hunk 6]
```diff
@@ -1144,7 +1269,7 @@ public void onEvent(PlayerEvent.DurationChanged event) {
 
     private void resetFlagsOnError() {
         isAdError = true;
-        adPlaybackCancelled = true;
+        //adPlaybackCancelled = true; //TODO: Manage state of this flag when waterfalling happens
```
Comment (giladna) [inline imaplugin/src/main/java/com/kaltura/playkit/plugins/ima/IMAPlugin.java]: what is this ?

File: imaplugin/src/main/java/com/kaltura/playkit/plugins/ima/IMAPlugin.java [hunk 7]
```diff
@@ -1282,7 +1412,11 @@ public void onAdEvent(com.google.ads.interactivemedia.v3.api.AdEvent adEvent) {
                                 preparePlayer(false);
                             }
                             if (getPlayerEngine() != null) {
-                                getPlayerEngine().play();
+                                if (isAdvertisingConfigLoading()) {
+                                    displayAd();
+                                } else {
+                                    getPlayerEngine().play();
+                                }
```
Comment (giladna) [inline imaplugin/src/main/java/com/kaltura/playkit/plugins/ima/IMAPlugin.java]: this code can be extracted to method

File: imaplugin/src/main/java/com/kaltura/playkit/plugins/ima/IMAPlugin.java (hunk 4 above)
Comment (GouravSna) [reply imaplugin/src/main/java/com/kaltura/playkit/plugins/ima/IMAPlugin.java]: Removed.

File: imaplugin/src/main/java/com/kaltura/playkit/plugins/ima/IMAPlugin.java [hunk 8]
```diff
@@ -137,6 +141,14 @@
     private PlayerEngineWrapper adsPlayerEngineWrapper;
     private Boolean playerPlayingBeforeAdArrived;
 
+    // Advertsing Configuration
+    private boolean isAdvertisingConfigured;
+    private IMAEventsListener imaEventsListener;
+    private boolean isAdvertisingConfigLoading;
+    private List<Long> advertisingConfigCuePoints;
+    private PKAdvertisingAdInfo pkAdvertisingAdInfo = null;
```
Comment (giladna) [inline imaplugin/src/main/java/com/kaltura/playkit/plugins/ima/IMAPlugin.java:148]: @GouravSna  why do we need  = null;

File: imaplugin/src/main/java/com/kaltura/playkit/plugins/ima/IMAPlugin.java [hunk 9]
```diff
@@ -1139,7 +1268,11 @@ public void onEvent(PlayerEvent.DurationChanged event) {
 
                             IMAPlugin.this.displayContent();
                             if (IMAPlugin.this.getPlayerEngine() != null) {
-                                IMAPlugin.this.getPlayerEngine().play();
+                                if (isAdvertisingConfigLoading()) {
```
Comment (GouravSna) [inline imaplugin/src/main/java/com/kaltura/playkit/plugins/ima/IMAPlugin.java]: Extract to a method

Comment (GouravSna) [reply imaplugin/src/main/java/com/kaltura/playkit/plugins/ima/IMAPlugin.java]: Fixed

**Review by giladna — APPROVED** [review]

**Review by giladna — APPROVED** [review]

**Pull request merged**